d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java

```diff
@@ -864,31 +974,33 @@ final class RpcRetryTask implements Runnable
     public void run()
     {
       startRpcStreamLocal();
-      for (ResourceType type : _resourceSubscribers.keySet())
+      for (ResourceType originalType : getResourceSubscribers().keySet())
       {
-        Set<String> resources = new HashSet<>(getResourceSubscriberMap(type).keySet());
-        if (resources.isEmpty() && getWildcardResourceSubscriber(type) == null)
+        Set<String> resources = new HashSet<>(getResourceSubscriberMap(originalType).keySet());
+        if (resources.isEmpty())
         {
           continue;
         }
-        ResourceType rewrittenType;
-        if (_subscribeToUriGlobCollection && type == ResourceType.D2_URI_MAP)
+
+        ResourceType adjustedType;
+        if (_subscribeToUriGlobCollection && originalType == ResourceType.D2_URI_MAP)
         {
           resources = resources.stream()
               .map(GlobCollectionUtils::globCollectionUrlForClusterResource)
               .collect(Collectors.toCollection(HashSet::new));
-          rewrittenType = ResourceType.D2_URI;
+          adjustedType = ResourceType.D2_URI;
         }
-        else
-        {
-          rewrittenType = type;
+        else {
+          adjustedType = originalType;
         }
-        // If there is a wildcard subscriber, we should always send a wildcard request to the server.
-        if (getWildcardResourceSubscriber(type) != null)
-        {
-          resources.add("*");
-        }
-        _adsStream.sendDiscoveryRequest(rewrittenType, resources);
+        _adsStream.sendDiscoveryRequest(adjustedType, resources);
+      }
+
+      for (ResourceType originalType: getWildcardResourceSubscribers().keySet())
+      {
+        ResourceType adjustedType = _subscribeToUriGlobCollection && originalType == ResourceType.D2_URI_MAP
+            ? ResourceType.D2_URI : originalType;
```

Comment:
So for example, this would now look like this:
```suggestion
        ResourceType adjustedType = isAdjustedUriGlobCollectionType(originalType)
            ? ResourceType.D2_URI : originalType;
```